### ISSUE TITLE
metadata.startup-script = "" triggers a change even if already = ""

### ISSUE THREAD

**Issue by billyfoss** [terraform-providers/terraform-provider-google#3687]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment
* If an issue is assigned to the "modular-magician" user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If an issue is assigned to a user, that user is claiming responsibility for the issue. If an issue is assigned to "hashibot", a community member has claimed the issue already.

### Terraform Version

[deployuser@40b184b5809b terraform]$ terraform -v
Terraform v0.11.11
+ provider.google v1.20.0
+ provider.template v2.1.1

Your version of Terraform is out of date! The latest version
is 0.11.14. You can update by downloading from www.terraform.io/downloads.html

### Affected Resource(s)

* google_compute_instance_from_template

### Terraform Configuration Files

```tf
resource "google_compute_instance_from_template" "instance" {
...
  metadata {
    startup-script = ""
  }
}

# Copy-paste your Terraform configurations here - for large Terraform configs,
# please use a service like Dropbox and share a link to the ZIP file. For
# security, you can also encrypt the files using our GPG public key: https://www.hashicorp.com/security
# If reproducing the bug involves modifying the config file (e.g., apply a config,
# change a value, apply the config again, see the bug) then please include both the
# version of the config before the change, and the version of the config after the change.
```

### Expected Behavior

No changes. Infrastructure is up-to-date.

This means that Terraform did not detect any differences between your
configuration and real physical resources that exist. As a result, no
actions need to be performed.

### Actual Behavior

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  ~ module.bastionserver01.google_compute_instance_from_template.public-instance
      metadata.%:              "1" => "2"
      metadata.startup-script: "" => <computed>

Plan: 0 to add, 1 to change, 0 to destroy.

### Steps to Reproduce

1. `terraform plan` - shows a change
1. `terraform apply` - applies the change
1. `terraform plan` - shows the same change

### References

https://github.com/terraform-providers/terraform-provider-google/blob/dab8fa8c8747f79a2a38db30e68361486745f0eb/google/resource_compute_instance.go#L790

https://github.com/terraform-providers/terraform-provider-google/pull/1081/files

The above merge works well to handle the issues around metadata_startup_script and metadata.startup-script.  We are just running into a special case when the metadata.startup-script is defined as "".  

As a workaround, setting metadata.startup-script = " " will apply and show the infrastructure as up-to-date. 

Also, I noticed that trying to update an existing startup-script to "" will see the state as already updated.  This will leave the startup script as not being updated. 

* #0000

**Comment by danawillow** [comment]
Hi @billyfoss, before I investigate I want to make sure I have your use case correct. Is the issue that the instance template defines a startup script, but you want to overwrite it to empty in the instance_from_template resource?

**Comment by billyfoss** [comment]
The use case is:
We have a module that creates instances with specifics like we want.  We default to a standard startup-script, but we allow the user to override the default.  When they override with "" to disable a startup script, the terraform plan never shows the infrastructure as up to date.  

I was able to recreate the issue with google_compute_instance, so it is not isolated to instance templates. 

Here is a small recreate.  

```
resource "google_compute_instance" "instance" {
  project = "my-project"
  name = "test-terraform-bug"
  zone = "us-east4-a"
  machine_type = "n1-standard-1"

  boot_disk {
    initialize_params {
      image = "centos-cloud/centos-7"
    }
  }

  network_interface {
    subnetwork = "https://www.googleapis.com/compute/v1/projects/my-project/regions/us-east4/subnetworks/my-network"
    network_ip = "10.0.0.9"
  }
  metadata {
    # Initially create the instance with a startup script
    #startup-script = "#!/bin/bash\necho Hello"
    # Update the script with an empty script
    startup-script = ""    
  }
}
```

Once the instance is created and updated with an empty startup script, I can run terraform apply and it keeps seeing the following. 
```
[deployuser@e6fc739f2390 terraform2]$ terraform apply
google_compute_instance.instance: Refreshing state... (ID: test-terraform-bug)

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  ~ google_compute_instance.instance
      metadata.%:              "0" => "1"
      metadata.startup-script: "" => ""


Plan: 0 to add, 1 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

google_compute_instance.instance: Modifying... (ID: test-terraform-bug)
  metadata.%:              "0" => "1"
  metadata.startup-script: "" => ""
google_compute_instance.instance: Still modifying... (ID: test-terraform-bug, 10s elapsed)
google_compute_instance.instance: Modifications complete after 12s (ID: test-terraform-bug)

Apply complete! Resources: 0 added, 1 changed, 0 destroyed.
```

**Comment by danawillow** [comment]
Just sent out a fix at https://github.com/GoogleCloudPlatform/magic-modules/pull/1816. I did notice that you're using version 1.20 of the provider, though- once merged, the fix will be in version 2.7.1 (or 2.8 if our next release is a minor version instead of a patch, which is more likely), so you'll need to upgrade in order to get it.

**Comment by billyfoss** [comment]
Thank you!  Upgrading the provider to 2.x is on our todo list.
